Conditional tab rendering in Playground

alcen opened the pull request:
In this commit, tabs in the Playground are now tied to their respective chapters. For example, Source §1 now only shows the "Introduction" tab.

A screenshot of what happens when 'Source §2' is selected in the Playground:

![condiiton-tabs](https://user-images.githubusercontent.com/49450743/62748469-502f2c00-ba8b-11e9-9cf4-7f1866a5b465.png)

This change is done in line with the future addition of the Substituter feature, which is only enabled for Source §2 and below - see #803

Note: disabling of tabs based on chapters is not done for Sourcecast and Sourcereel as I am not sure how it would affect the recording. If the developers for these features would advise how to do so, that would be great!

Note 2: this Pull Request was originally full of different unrelated changes for Playground side content. The changes have been since rebased into their own PRs:

| Description                               | Pull Request No.    |
|---------------------------------------|------|
| Assessments default tab fix     | #819 |
| Conditional tab rendering       | #808 |
| Default tab text                | #820 |
| List visualiser scrolling fix   | #817 |
| List visualiser orange flashing | #818 |

&nbsp;

Thanks to @lumos309 for a bug fix for a Playground crash when the editor was trying to find the non-existent Inspector/Env Visualiser tabs!

alcen commented:
For reference, these are the raw default texts used in the code:

Data Visualiser:
```
The data visualizer visualises data structures.
<br/>
<br/>
It is activated by calling the function <code>draw_data(the_data)</code>,
where <code>the_data</code> would be the data structure that you want to visualise.
```

Inspector:
```
The inspector generates a list of variable bindings based on breakpoints set in the editor.
<br/>
<br/>
It is activated by clicking on the gutter of the editor (where all the line numbers are,
on the left) to set a breakpoint, and then running the program. Alternatively, you may
add the line <code>debugger;</code> to your program!
```

Env Visualiser:
```
The environmental model visualizer generates the environmental model diagram
based on breakpoints set in the editor.
<br/>
<br/>
It is activated by clicking on the gutter of the editor (where all the line numbers are,
on the left) to set a breakpoint, and then running the program. Alternatively, you may
add the line <code>debugger;</code> to your program!
```

martin-henz commented:
> For reference, these are the raw default texts used in the code:
> 
> Data Visualiser:
> 
> ```
> The data visualizer visualises data structures.
> <br/>
> <br/>
> It is activated by calling the function <code>draw_data(the_data)</code>,
> where <code>the_data</code> would be the data structure that you want to visualise.
> ```
> 
> Inspector:
> 
> ```
> The inspector generates a list of variable bindings based on breakpoints set in the editor.
> <br/>
> <br/>
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. Alternatively, you may
> add the line <code>debugger;</code> to your program!
> ```
> 
> Env Visualiser:
> 
> ```
> The environmental model visualizer generates the environmental model diagram
> based on breakpoints set in the editor.
> <br/>
> <br/>
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. Alternatively, you may
> add the line <code>debugger;</code> to your program!
> ```

Note that gutter click is different from debugger; statements. We decided not to support debugger; statements, after discussing with @podocarp . In fact, this Issue requests removing support of debugger; statements in Source: https://github.com/source-academy/js-slang/issues/276

So let's not mention debugger; in the default texts.

alcen commented:
> For reference, these are the raw default texts used in the code:
> Data Visualiser:
> The data visualizer visualises data structures.
> <br/>
> <br/>
> It is activated by calling the function <code>draw_data(the_data)</code>,
> where <code>the_data</code> would be the data structure that you want to visualise.
> 
> Inspector:
> The inspector generates a list of variable bindings based on breakpoints set in the editor.
> <br/>
> <br/>
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. Alternatively, you may
> add the line <code>debugger;</code> to your program!
> 
> Env Visualiser:
> The environmental model visualizer generates the environmental model diagram
> based on breakpoints set in the editor.
> <br/>
> <br/>
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. Alternatively, you may
> add the line <code>debugger;</code> to your program!
> 
> 
> Note that gutter click is different from debugger; statements. We decided not to support debugger; statements, after discussing with @podocarp . In fact, this Issue requests removing support of debugger; statements in Source: source-academy/js-slang#276
> So let's not mention debugger; in the default texts.

Oh, I see. Okay, will change this!

martin-henz commented:
Here are my suggestions for the text:

Data Visualizer:

The data visualizer visualises data structures.
<br/>
<br/>
It is activated by calling the function <code>draw_data(the_data)</code>,
where <code>the_data</code> would be the data structure that you want to visualise.
The data visualizer uses box-and-pointer diagrams, as introduced in 
Structure and Interpretation of Computer Programs, JavaScript Adaptation, Chapter 2.

Inspector:

The inspector generates a list of variable bindings based on breakpoints set in the editor.
<br/>
<br/>
It is activated by clicking on the gutter of the editor (where all the line numbers are,
on the left) to set a breakpoint, and then running the program. 

Environment Visualizer:

The environment visualizer generates the environment model diagram
based on breakpoints set in the editor.
<br/>
<br/>
It is activated by clicking on the gutter of the editor (where all the line numbers are,
on the left) to set a breakpoint, and then running the program. The environment model diagram follows a notation introduced in Structure and Interpretation of Computer Programs, JavaScript Adaptation, Section 3.2.

martin-henz commented:
> Here are my suggestions for the text:
> 
> Data Visualizer:
> 
> The data visualizer visualises data structures.
> 
> 
> 
> 
> It is activated by calling the function `draw_data(the_data)`,
> where `the_data` would be the data structure that you want to visualise.
> The data visualizer uses box-and-pointer diagrams, as introduced in
> Structure and Interpretation of Computer Programs, JavaScript Adaptation, Chapter 2.
> 
> Inspector:
> 
> The inspector generates a list of variable bindings based on breakpoints set in the editor.
> 
> 
> 
> 
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program.
> 
> Environment Visualizer:
> 
> The environment visualizer generates the environment model diagram
> based on breakpoints set in the editor.
> 
> 
> 
> 
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. The environment model diagram follows a notation introduced in Structure and Interpretation of Computer Programs, JavaScript Adaptation, Section 3.2.

It would be good to use a link to the textbook. Note that the textbook URL is in `cadet-frontend/src/utils/constants.ts`

lumos309 commented:
Hmm, this seems to break for me when running any program with Source 1 or 2 selected. The inspector code still tries to run and update a non-existent tab, which crashes the whole thing.

alcen commented:
> Here are my suggestions for the text:
> Data Visualizer:
> The data visualizer visualises data structures.
> It is activated by calling the function draw_data(the_data),
> where the_data would be the data structure that you want to visualise.
> The data visualizer uses box-and-pointer diagrams, as introduced in
> Structure and Interpretation of Computer Programs, JavaScript Adaptation, Chapter 2.
> Inspector:
> The inspector generates a list of variable bindings based on breakpoints set in the editor.
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program.
> Environment Visualizer:
> The environment visualizer generates the environment model diagram
> based on breakpoints set in the editor.
> It is activated by clicking on the gutter of the editor (where all the line numbers are,
> on the left) to set a breakpoint, and then running the program. The environment model diagram follows a notation introduced in Structure and Interpretation of Computer Programs, JavaScript Adaptation, Section 3.2.
> 
> It would be good to use a link to the textbook. Note that the textbook URL is in cadet-frontend/src/utils/constants.ts

Ok, I'll add this in!



> Hmm, this seems to break for me when running any program with Source 1 or 2 selected. The inspector code still tries to run and update a non-existent tab, which crashes the whole thing.

Will look into this!

geshuming commented on src/sagas/workspaces.ts:
```diff
@@ -357,7 +357,7 @@ export function* evalCode(
     (actionType === actionTypes.EVAL_EDITOR || actionType === actionTypes.DEBUG_RESUME) &&
     context.chapter > 2;
-  if (!context.runtime.debuggerOn) {
+  if (!context.runtime.debuggerOn && context.chapter > 2) {
```
Can we make this more logical, this part has always been making my brain explode.

```javascript
const useDebugger = actionType !== (actionTypes.EVAL_EDITOR || actionType !== actionTypes.DEBUG_RESUME) && context.chapter > 2;

if (useDebugger) {
  inspectorUpdate(undefined);
}
```

Honestly though, I don't even think the above code segment makes any sense. Please comment on what this code segment is even intended to do.

geshuming replied:
Okay, `debuggerOn` is used as a flag for slang.

So the correct code should look like 

```javascript
context.runtime.debuggerOn = actionType !== actionTypes.EVAL_EDITOR && actionType !== actionTypes.DEBUG_RESUME && context.chapter > 2;

if (context.runtime.debuggerOn) {
  // Resets the inspector to prepare for new output
  inspectorUpdate(undefined);
}
```

lumos309 replied:
This was done by the INSPECT team, so I'm not certain about how it works either. Having said that, this:

> Okay, `debuggerOn` is used as a flag for slang.
> 
> So the correct code should look like
> 
> ```js
> context.runtime.debuggerOn = (actionType === actionTypes.EVAL_EDITOR || actionType === actionTypes.DEBUG_RESUME) && context.chapter > 2;
> 
> if (context.runtime.debuggerOn) {
>   // Resets the inspector to prepare for new output
>   inspectorUpdate(undefined);
> }
> ```

doesn't seem to be the same boolean expression as the original code. `!(actionType === actionTypes.EVAL_EDITOR || actionType === actionTypes.DEBUG_RESUME) && context.chapter > 2` will call `inspectorUpdate(undefined)` in the original code, but not in the suggested amendment.

geshuming replied:
My mistake. The clause should be 

```javascript
if (!context.runtime.debuggerOn)
``` 

just like the original code

lumos309 replied:
No, this won't work because it can trigger `inspectedUpdate(undefined)` even when `!(context.chapter > 2)`, and that crashes the entire webpage. That's what the change was originally for.

geshuming replied:
Ok, can we clean this up then. The logic here is very confusing. Can we add comments to clarify?

alcen commented:
> Can we split this PR into smaller PRs, all rebased. The PR title doesn't even match what this PR is trying to achieve.

Ok, I have split it up. The new pull requests can be found at:

Assessments default tab fix: #819
Conditional tab rendering: #808
Default tab text: #820
List visualiser scrolling fix: #817 
List visualiser orange flashing: #818
